File: 28.find-the-index-of-the-first-occurrence-in-a-string.go
/*
 * @lc app=leetcode id=28 lang=golang
 *
 * [28] Find the Index of the First Occurrence in a String
 */

//haystack = "sadbutsad", needle = "sad"
// Output:0

// string[start:end]
// start：起始索引（包含此索引）。
// end：結束索引（不包含此索引）。
// 結果：從 start 到 end-1 的子字串。

// @lc code=start
func strStr(haystack string, needle string) int {

	if len(needle) == 0 || len(haystack) == 0 {
		return 0
	}
	m, n := len(needle), len(haystack)

	for i := 0; i <= n-m; i++ {
		if haystack[i:i+m] == needle {
			return i
		}
	}
	return -1

	// if len(needle) == 0 {
	// 	return 0
	// }

	// m, n := len(needle), len(haystack)

	// // Sliding window approach
	// //n-m 為了限制不要超出比較的長度,不要超出needle的長度
	// //haystack = "leetcode", needle = "leeto"
	// // i ~ i+m 每移動一次,就跟著i+m

	// for i := 0; i <= n-m; i++ {
	// 	if haystack[i:i+m] == needle {
	// 		return i
	// 	}
	// }

	// return -1
}

// @lc code=end

File: 28.find-the-index-of-the-first-occurrence-in-a-string.py
#
# @lc app=leetcode id=28 lang=python
#
# [28] Find the Index of the First Occurrence in a String
#
#Input: haystack = "sadbutsad", needle = "sad"
#Output:0
# @lc code=start
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        m, n = len(haystack), len(needle)
        if n ==0:
            return 0 #Empty needle is always found at index 0

        # Sliding window
        # 為了不超出needle的長度,最多只循環到
        # range(8 - 5 + 1) = range(4)，即索引 0 到 3
        # To stay within the bounds of needle, loop up to range(8 -5 + 1), range(4), which covers index 0 to 3
        for i in range(m-n+1):
            # 比較子字串的長度與needle是否相等
            #Compare substring of length n with needle
            # leetcode -> haystack[0:5] = "leetc" -> index 0~4
            if haystack[i:i + n] ==needle:
                return i

        return -1
        

# @lc code=end

